transport/http/agent_test.go

```diff
@@ -0,0 +1,389 @@
+package http
+
+import (
+	"testing"
+
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/internal/errors"
+	"go.uber.org/yarpc/transport/peer/hostport"
+	"go.uber.org/yarpc/transport/transporttest"
+
+	"github.com/crossdock/crossdock-go/assert"
+	"github.com/golang/mock/gomock"
+)
+
+type peerExpectation struct {
+	identifier  hostport.PeerIdentifier
+	subscribers []*transporttest.MockPeerSubscriber
+}
+
+func createPeerExpectations(
+	mockCtrl *gomock.Controller,
+	hostports []string,
+	subscribers int,
+) []peerExpectation {
+	expectations := make([]peerExpectation, 0, len(hostports))
+	for _, hp := range hostports {
+		hpid := hostport.NewPeerIdentifier(hp)
+		subs := make([]*transporttest.MockPeerSubscriber, 0, subscribers)
+		for i := 0; i < subscribers; i++ {
+			subs = append(subs, transporttest.NewMockPeerSubscriber(mockCtrl))
+		}
+		expectations = append(expectations, peerExpectation{
+			identifier:  hpid,
+			subscribers: subs,
+		})
+	}
+	return expectations
+}
+
+// peerIdentifierMatcher allows us to compare peerIdentifiers with Peers through gomock
+type peerIdentifierMatcher hostport.PeerIdentifier
+
+// Matches returns whether x is a match.
+func (pim peerIdentifierMatcher) Matches(x interface{}) bool {
+	res, ok := x.(transport.PeerIdentifier)
+	if !ok {
+		return false
+	}
+
+	return res.Identifier() == hostport.PeerIdentifier(pim).Identifier()
+}
+
+// String describes what the matcher matches.
+func (pim peerIdentifierMatcher) String() string {
+	return string(hostport.PeerIdentifier(pim).Identifier())
+}
+
+func TestAgent(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	type testStruct struct {
+		msg           string
+		agent         *Agent
+		appliedFunc   func(*Agent) error
+		expectedPeers []peerExpectation
+		expectedErr   error
+	}
+	tests := []testStruct{
+		func() (s testStruct) {
+			s.msg = "one retain"
+			s.agent = NewDefaultAgent()
+			s.expectedPeers = createPeerExpectations(
+				mockCtrl,
+				[]string{"localhost:1234"},
+				1,
+			)
+			s.appliedFunc = func(a *Agent) error {
+				_, err := a.RetainPeer(s.expectedPeers[0].identifier, s.expectedPeers[0].subscribers[0])
```

Comment:
The way these tests are written is really confusing. The code that drives the
test interacts with the expectations so it's not clear where the boundary
between input and expected output is.

Things you can do to make this clearer:

-   Use local variables for mock subscribers rather than
    `s.expectedPeers[x].subscribers[y]`.
    
    ```go
    sub := NewMockPeerSubscriber(mockCtrl)
    
    // ...
    
    s.appliedFunc = func(a *Agent) error {
        _, err := a.RetainPeer(..., sub)
    }
    ```
-   Pull error cases into a separate test function, reducing noise from
    expectedErr cases. That will allow `appliedFunc` to never return `error`
    and fail the test early if it does run into an error.
-   Hard-code peer names rather than accessing expectation-side of things:

    ```go
    s.appliedFunc = func(a *Agent) error {
        _, err := a.RetainPeer("localhost:1234", ...)
    }
    ```
    
----

Optionally: Just throwing an idea out there which I imagine will make the test
cases clearer

```
// Something like,
func() (s testStruct) {
    sub1 := NewMockSubscriber(mockCtrl)
    sub2 := NewMockSubscriber(mockCtrl)

    s.msg = "one retain, one release"
    s.actions = []action{
        retain{peer: "localhost:1234", subscriber: sub1},
        retain{peer: "localhost:1234", subscriber: sub2},
        release{peer: "localhost:1234", subscriber: sub1}
    }
    s.wantPeers = expectedPeers{
        "localhost:1234": {sub2},
    }
    return
}
```

Note that with this approach, if you pull out errors into a separate test
function, it will allow your test driver to always assert that the retain/
release actions never fail in these test cases. Further, you will be able to
execute these actions concurrently to test your thread-safety further.